source/extensions/common/tap/tap_matcher.h

```diff
@@ -24,20 +24,26 @@ using MatcherPtr = std::unique_ptr<Matcher>;
  * - In order to make this computationally efficient, the matching tree is kept in a vector, with
  *   all references to other matchers implemented using an index into the vector. The vector is
  *   effectively a preorder traversal flattened N-ary tree.
- * - The previous point allows the creation of a per-stream/request vector of booleans of the same
- *   size as the matcher vector. Then, when match status is updated given new information, the
- *   vector of booleans can be easily updated using the same indexes as in the constant match
- *   configuration.
+ * - The previous point allows the creation of a per-stream/request vector of match statuses of
+ *   the same size as the matcher vector. Then, when match status is updated given new
+ *   information, the vector of match statuses can be easily updated using the same indexes as in
+ *   the constant match configuration.
  * - Finally, a matches() function can be trivially implemented by looking in the status vector at
  *   the index position that the current matcher is located in.
- *
- * TODO(mattklein123): Currently, any match updates perform a recursive call on any child match
- * nodes. It's possible that we can short circuit this in certain cases but this needs more
- * thinking (e.g., if an OR matcher already has one match and it's not possible for a matcher to
- * flip from true to false).
  */
 class Matcher {
 public:
+  struct MatchStatus {
+    bool operator==(const MatchStatus& rhs) const {
+      return matches_ == rhs.matches_ && might_change_status_ == rhs.might_change_status_;
+    }
+
+    bool matches_{false};            // Does the matcher currently match?
+    bool might_change_status_{true}; // Is it possible for matches_ to change in subsequent updates?
```

Comment:
Can you split this to a separate PR? This looks like it could be cleanly done standalone. Thanks.

Reply:
OK